reps/2023-06-06-simplify-sync.md

```diff
@@ -0,0 +1,126 @@
+
+Third, with the growth of LLMs and other large models, we are seeing larger and larger checkpoints being used. This motivates spending more effort in cleaning up and hardening this critical path, and it may cause more pain for LLM users than for users for smaller models. Simplifying the design here also enables us to extend support for more advanced features such as multi-rank checkpointing withour incurring excessive technical debt.
+
+These three pain points have grown significantly over the past year, based both on user and internal feedback. This shouldn't be too surprising, as the persistent storage APIs have never been significantly revised before, beyond introduction of the new `storage_path` option, and similarly the Checkpoint API is in beta.
+
+### Should this change be within `ray` or outside?
+
+main `ray` project. Changes are made to Ray Train and Tune.
+
+## Stewardship
+
+### Required Reviewers
+The proposal will be open to the public, but please suggest a few experienced Ray contributors in this technical domain whose comments will help this proposal. Ideally, the list should include Ray committers.
+
+@matthewdeng, @krfricke
+
+### Shepherd of the Proposal (should be a senior committer)
+To make the review process more productive, the owner of each proposal should identify a **shepherd** (should be a senior Ray committer). The shepherd is responsible for working with the owner and making sure the proposal is in good shape (with necessary information) before marking it as ready for broader review.
+
+@pcmoritz
+
+## User Story
+
+Persistence will center around the recently introduced `storage_path` option.
+
+### Persistent trial directory
+
+Ray Train provides the abstraction of a persistent trial directory. Each trial has its current working directory set to a trial specific directory (e.g., `~/ray_results/experiment_1/trial_1`). Train will synchronize this trial directory with persistent storage specified via the `storage_path` option.
+
+This directory is persistent since Train is syncing it to persistent storage for the user. The experiment driver will record experiment results and other metadata to a similar directory on the head node, which will also synchronize to persistent storage.
+
+![overview](user-story.png)
+
+### Checkpoints
+
+While trials can write arbitrary data to the persistent trial directory, they cannot be guaranteed they can read it back on checkpoint/restore. Data that is needed for resume must be recorded via the `train.report(checkpoint)` API.
+
+For example, this is how you could record a Torch checkpoint:
+
+```python
+   torch.save(model, "path/to/dir")
+   train.report(checkpoint=Checkpoint.from_directory("path/to/dir"))
```

Comment:
Just confirming, will there be a change to the structure in which this is persisted compared to the current?

In the current implementation, the _contents_ of the checkpoint directory will be moved to a `checkpoint_{id}` directory. If I remember correctly, with this proposal we would want to store the opaque checkpoint dir/file in the `checkpoint_{id}` directory. Is that right?

Original directory:
```
path/to/dir/model.pt
```
Current:
```
{storage_path}/{experiment}/{trial}/checkpoint_{id}/model.pt
```

Proposal?
```
{storage_path}/{experiment}/{trial}/checkpoint_{id}/dir
```

Reply:
There is no change, the current is the same as the new proposal behavior.

Reply:
Ah the current implementation will "rename" the directory to `checkpoint_{id}`. See this naive example:

```python
import os
from ray.train.torch import TorchTrainer
from ray.air import ScalingConfig, session, Checkpoint

def train_func():
    checkpoint_dir = "checkpoint_dir"
    os.makedirs(checkpoint_dir, exist_ok=True)
    open(f"{checkpoint_dir}/checkpoint_file1", "w").close()
    open(f"{checkpoint_dir}/checkpoint_file2", "w").close()
    session.report({}, checkpoint=Checkpoint.from_directory(checkpoint_dir))

trainer = TorchTrainer(train_loop_per_worker=train_func, scaling_config=ScalingConfig(num_workers=2))
trainer.fit()
```

Result:
```
├── TorchTrainer_70610_00000_0_2023-07-17_16-48-33
│   ├── checkpoint_000000
│   │   ├── _preprocessor
│   │   ├── checkpoint_file1
│   │   └── checkpoint_file2
```

Note that there is no `checkpoint_dir`  in `checkpoint_000000`.

In the proposal did we want to move to something like this?

```
├── TorchTrainer_70610_00000_0_2023-07-17_16-48-33
│   ├── checkpoint_000000
│   │   ├── _metadata
│   │   ├── checkpoint_dir
│   │   │   ├── checkpoint_file1
│   │   │   └── checkpoint_file2
```

Reply:
I don't think so--- if you look at the prototype we don't have nested dirs created. The data is directly under `checkpoint_000000`.

Reply:
Oh I see. We previously discussed supporting creating a checkpoint directly from a single file (as opposed to directory), should we still support that?

Reply:
I do not think we should support that.